src/tree.rs

```diff
@@ -135,3 +138,159 @@ pub fn minimum_spanning_tree(
+            let pop = pops[node.index()];
+
+            // todo: factor out expensive clones
+            // Mark as seen; push to queue if only one unseen neighbor
+            let unseen_neighbors: Vec<NodeIndex> = spanning_tree
+                .graph
+                .neighbors(node)
+                .filter(|node| !seen_nodes[node.index()])
+                .collect();
+            // eprintln!("unseen_neighbors: {}", unseen_neighbors.len());
+            if unseen_neighbors.len() == 1 {
+                // this will be false if root
+                let neighbor = unseen_neighbors[0];
+                pops[neighbor.index()] += pop.clone();
+                let mut current_partition_tracker = same_partition_tracker[node.index()].clone();
+                same_partition_tracker[neighbor.index()].append(&mut current_partition_tracker);
+                // eprintln!("node pushed to queue (pop = {}, target = {}): {}", pops[neighbor.index()], pop_target, neighbor.index());
+
+                if !node_queue.contains(&neighbor) {
+                    node_queue.push_back(neighbor);
+                }
+            } else if unseen_neighbors.len() == 0 {
+                break;
+            } else {
+                continue;
+            }
+            // pops[node.index()] = 0.0; // not needed?
+
+            // Check if balanced
+            if pop >= pop_target * (1.0 - epsilon) && pop <= pop_target * (1.0 + epsilon) {
+                // slightly different
+                // eprintln!("balanced node found: {}", node.index());
+                balanced_nodes.push((node.index(), same_partition_tracker[node.index()].clone()));
+            }
+
+            seen_nodes[node.index()] = true;
+        }
+        balanced_nodes
+    });
+
+    Ok(balanced_nodes)
+}
+
+/// Bipartition graph into two contiguous, population-balanced components.
+/// Assumes that graph is contiguous.
+///
+/// :param PyGraph graph: Undirected graph
+/// :param weight_fn: A callable object (function, lambda, etc) which
+///     will be passed the edge object and expected to return a ``float``. This
+///     tells retworkx/rust how to extract a numerical weight as a ``float``
+///     for edge object. Some simple examples are::
+///
+///         minimum_spanning_tree(graph, weight_fn: lambda x: 1)
+///
+///     to return a weight of 1 for all edges. Also::
+///
+///         minimum_spanning_tree(graph, weight_fn: float)
+///
+///     to cast the edge object as a float as the weight.
+/// :param float default_weight: If ``weight_fn`` isn't specified this optional
+///     float value will be used for the weight/cost of each edge.
+///
+/// :returns: A set of nodes in one half of the spanning tree
+///
+#[pyfunction]
+#[pyo3(text_signature = "(graph, weight_fn, pop, target_pop, epsilon)")]
+pub fn bipartition_tree(
+    py: Python,
+    graph: &graph::PyGraph,
+    weight_fn: PyObject,
+    py_pops: &PyList,
+    py_pop_target: &PyFloat,
+    py_epsilon: &PyFloat,
```

Comment:
For these you can let the macro handle the conversion from python to rust for us. Something like:

```suggestion
    py_pops: Vec<f64>,
    py_pop_target: f64,
    py_epsilon: f64,
```

Then we can avoid the unsafe and gil manipulation in the code here and just work purely in rust. I'd probably rename all the args here to drop the `py_` prefix.